providers/file/provider.go

```diff
@@ -0,0 +1,292 @@
+limitations under the License.
+*/
+
+package file
+
+import (
+	"context"
+	"fmt"
+	"maps"
+	"os"
+	"path/filepath"
+	"slices"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/google/go-cmp/cmp"
+	"gopkg.in/fsnotify.v1"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	mcmanager "sigs.k8s.io/multicluster-runtime/pkg/manager"
+	"sigs.k8s.io/multicluster-runtime/pkg/multicluster"
+)
+
+var _ multicluster.Provider = &Provider{}
+
+// Options defines the options for the file-based cluster provider.
+type Options struct {
+	// KubeconfigFiles are paths to kubeconfig files.
+	// The default depends on the KubeconfigDirs variable.
+	KubeconfigFiles []string
+
+	// KubeconfigDirs are directories to search for kubeconfig files matching the
+	// globs specified in KubeconfigGlobs.
+	//
+	// If either one or both of KubeconfigFiles or KubeconfigDirs are
+	// set both are used as input for the provider.
+	// If both are empty defaults are applied in order of precedence:
+	// 1. If the KUBECONFIG environment variable is set and contains
+	//       a path to a valid file it is used in KubeconfigFiles.
+	// 2. If ~/.kube/config exists it is used in KubeconfigFiles.
+	// 3. The working directory is used in KubeconfigDirs.
+	KubeconfigDirs []string
+
+	// KubeconfigGlobs are the glob patterns to match kubeconfig files
+	// in directories.
+	// Default is DefaultKubeconfigGlobs.
+	KubeconfigGlobs []string
+}
+
+// DefaultKubeconfigGlobs are the default glob patterns when searching
+// for kubeconfig files in a directory.
+var DefaultKubeconfigGlobs = []string{
+	"kubeconfig.yaml",
+	"kubeconfig.yml",
+	"*.kubeconfig",
+	"*.kubeconfig.yaml",
+	"*.kubeconfig.yml",
+}
+
+func (p *Provider) defaultKubeconfigPaths() ([]string, []string) {
+	if envKubeconfig := os.Getenv("KUBECONFIG"); envKubeconfig != "" {
+		return []string{envKubeconfig}, []string{}
+	}
+
+	defaultKubeconfig := os.ExpandEnv("$HOME/.kube/config")
+	if _, err := os.Stat(defaultKubeconfig); err == nil {
+		return []string{defaultKubeconfig}, []string{}
+	}
+
+	pwd, err := os.Getwd()
+	if err != nil {
+		p.log.Error(err, "error getting working directory, defaulting to '.'")
+		pwd = "."
+	}
+
+	return []string{}, []string{pwd}
+}
+
+// New returns a new Provider with the given options.
+func New(opts Options) (*Provider, error) {
+	p := new(Provider)
+	p.opts = opts
+
+	if len(p.opts.KubeconfigFiles) == 0 && len(p.opts.KubeconfigDirs) == 0 {
+		p.opts.KubeconfigFiles, p.opts.KubeconfigDirs = p.defaultKubeconfigPaths()
+	}
+
+	if len(p.opts.KubeconfigGlobs) == 0 {
+		p.opts.KubeconfigGlobs = DefaultKubeconfigGlobs
+	}
+
+	p.log = log.Log.WithName("file-cluster-provider")
+	p.clusters = make(map[string]cluster.Cluster)
+	p.clusterCancel = make(map[string]func())
+
+	fmt.Printf("file cluster provider initialized %q %q\n", p.opts.KubeconfigFiles, p.opts.KubeconfigDirs)
```

Comment:
Leftover from debugging? Should this be a `p.log` call?

Reply:
Yeah, leftover from debugging - but i think this would make sense to leave as a log statement, might help with debugging